Make it possible to reduce available patterns

maxmeyer opened the pull request:
This PR is based on rspec-PR and testsuite-PR.

This PR makes it possible to reduce the available patterns. To make this possible it moves code from the `PatternGenerator`-class to a separate class I named `PatternDb`. This class handles all the checks etc. To make the library more testable I also moved some code to helpers methods I can also use in the tests.

maxmeyer commented:
I also added `aruba` to make filesystem-tests possible though I didn't add tests which make use of it.

ttaylorr commented on lib/geo_pattern/pattern_db.rb:
```diff
@@ -0,0 +1,70 @@
+    public
+
+    def initialize
+      @db = {
```
`@patterns` might be a better name, considering it is under an `attr_reader`.

ttaylorr commented on lib/geo_pattern/pattern_db.rb:
```diff
@@ -0,0 +1,70 @@
+      output_warning = false
+      abort("Error: At least one of the requested patterns \"#{patterns.join(", ")}\" is invalid") unless valid?(patterns)
+
+      result = patterns.map do |p|
```
This is a bit pedantic, but `pattern` would really do better here than `p`.

maxmeyer commented:
Ok. Fixed.

maxmeyer commented:
Rebased.

maxmeyer commented:
Ready to go :rocket:

jasonlong commented:
Thanks @maxmeyer. I'm going to take a little time to get my head around this one. Hopefully I can do so in the next couple of days. Thanks again for all the work here!

maxmeyer commented:
No problem. You're welcome. Thanks a lot for making this library public. :smiley: 

> Hopefully I can do so in the next couple of days. 

Would be great! I started using this library in a few projects of mine to generate some small eyecatchers. And spread the word about it. It's quite amazing how beautiful those generated images are.

:shipit:

ttaylorr commented:
Glad we're shipping these so quickly!  :shipit:

On Thu, Jan 08, 2015 at 12:59:59PM -0800, maxmeyer wrote:
> No problem. You're welcome. Thanks a lot for making this library public. :smiley:
>
> > Hopefully I can do so in the next couple of days.
>
> Would be great! I started using this library in a few projects of mine to generate some small eyecatchers. And spread the word about it. It's quite amazing how beautiful those generated images are.
>
> :shipit:
>
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/jasonlong/geo_pattern/pull/40#issuecomment-69248211

maxmeyer commented:
BTW:

I added a few more tests. While writing them I also found a bug in the previous implementation of mine. Which is fixed now. I also added a proper exception for an invalid pattern which will be raised instead of "SystemExit" which is used by `abort`. Something I would expect to be raise by a commandline application but not a library.

maxmeyer commented on lib/geo_pattern/errors.rb:
```diff
@@ -0,0 +1,7 @@
+module GeoPattern
+  # user errors
+  class UserError < StandardError; end
```
I think using this kind of pattern for exceptions is a question of style. If you do not like it, I can remove the "intermediate" error.

jasonlong replied:
I don't have an opinion on this. @ttaylorr, do you?

ttaylorr replied:
I use this pattern all the time -- looks great.

maxmeyer commented on lib/geo_pattern/pattern_db.rb:
```diff
@@ -0,0 +1,82 @@
+        end
+      end
+
+      puts SVG.as_comment("String pattern references are deprecated as of 1.3.0") if output_warning
```
Why did you output the deprecation warning as "SVG-comment"? I reused this line from the previous implementation.

jasonlong replied:
The deprecation warning is output w/ the SVG string as a comment so that the resulting SVG is still valid. I'm not sure many people will see the warning this way, but this seems like the best way since this is the only output.

maxmeyer replied:
But `puts` prints the comment on STDOUT. And it should be normally not part of the SVG.

jasonlong commented:
The changes here seem ok to me. Much of it effects @ttaylorr's recent restructuring changes, so I want to make sure we have a :thumbsup: from him as well.

maxmeyer commented:
Ok. Thanks for the feedback.

maxmeyer commented:
@ttaylorr ping :-)

ttaylorr commented:
Sorry that took so long, it's been a crazy few days for me!

Everything looks fine to me, but I would've implemented a `PatternSieve` class which has the responsibility of selecting available filters.  If we don't want to go that route, this PR is fine to merge as-is.

maxmeyer commented:
I'm not sure if I understand you correctly. Do you suggest to move this [method](https://github.com/jasonlong/geo_pattern/pull/40/files#diff-e5549ea50e5e5f867be690a7fc2d1c5fR38) and those following [this](https://github.com/jasonlong/geo_pattern/pull/40/files#diff-e5549ea50e5e5f867be690a7fc2d1c5fR63) to a separate class? I'm open for suggestions. :-)

jasonlong commented:
Friendly @ttaylorr ping. You aren't busy are you? :wink:

maxmeyer commented:
:-)

maxmeyer commented:
I think I've got an idea what he has got in his mind. @ttaylorr I think we're mixing multiple things in `PatternDb`:

1. "Translating" strings to classes
2. Checking if a pattern is valid
3. Filtering patterns

Correct @ttaylorr? So I will try to splitup `PatternDb`into two or three classes. One for each task mentioned above.

ttaylorr commented:
Yup.  `PatternDB` has too many responsibilities right now.  I think the
`PatternDB` class is fine to take those first two tasks, but the filtering
of `Pattern`s should be done somewhere else.

What I had suggested earlier (albeit in some fairly ambiguous language) was
to create a `PatternSieve` which has a `PatternDB` as reference and a
`Filter` upon which it can narrow down options in the backing DB.  A call
to `PatternSieve#valid_patterns` should return all elements in the backing
`PatternDB` `#select`ed with the Filter object.

The Filter class would have one method `Filter#is_valid(Pattern)` which
would take in a pattern and return true/false depending on whether or not
that pattern matches the criteria set by that filter.

Make sense?

On Sat, Jan 24, 2015 at 3:06 PM, maxmeyer <notifications@github.com> wrote:

> I think I've got an idea what he has got in his mind. @ttaylorr
> <https://github.com/ttaylorr> I think we're mixing multiple things in
> PatternDb:
>
>    1. "Translating" strings to classes
>    2. Checking if a pattern is valid
>    3. Filtering patterns
>
> Correct @ttaylorr <https://github.com/ttaylorr>? So I will try to splitup
> PatternDbinto two or three classes. One for each task mentioned above.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/jasonlong/geo_pattern/pull/40#issuecomment-71334914>.
>



-- 
Thanks,
Taylor Blau

ttaylorr.com

maxmeyer commented:
Yes. I put together some more commits and push them as soon as I can. :grin: I think
`is_valid` -> `valid?` is a better choice ;-P.

ttaylorr commented:
Mixing my Java conventions with my Ruby conventions!  `Filter#valid?` is a
much better choice in Ruby :)

On Sat, Jan 24, 2015 at 3:12 PM, maxmeyer <notifications@github.com> wrote:

> Yes. I put together some more commits and push them as soon as I can. [image:
> :grin:] I think
> is_valid -> valid? is a better choice ;-P.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/jasonlong/geo_pattern/pull/40#issuecomment-71335189>.
>



-- 
Thanks,
Taylor Blau

ttaylorr.com

maxmeyer commented:
Maybe I rename `PatternDB` to `PatternTranslator`. But only if I also outsource the validation part.

ttaylorr commented:
I would go with `PatternStore`, but either will work :smile:.

On Sat, Jan 24, 2015 at 3:14 PM, maxmeyer <notifications@github.com> wrote:

> Maybe I rename PatternDB to PatternTranslator. But only if I also
> outsource the validation part.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/jasonlong/geo_pattern/pull/40#issuecomment-71335244>.
>



-- 
Thanks,
Taylor Blau

ttaylorr.com

maxmeyer commented:
@ttaylorr 
I like the name `PatternStore`. Changed the implementation already. I think I will push the new implementation this evening.

maxmeyer commented:
One thing I would like you to reconsider:
Changing the "pattern:"-option one last time from support classes to supporting `Symbols` by default. I think classes reveal too much implementation details and Symbols are much more common. The new implementaiton will support them as well. I will just change the deprecation warnings.

ttaylorr commented:
I disagree with it revealing too many implementation details, but for the
sake of `Symbol`s being more common, this seems like a win to me.

On Sun, Jan 25, 2015 at 1:25 PM, maxmeyer <notifications@github.com> wrote:

> One thing I would like you to reconsider:
> Changing the "pattern:"-option one last time from support classes to
> supporting Symbols by default. I think classes reveal too much
> implementation details and Symbols are much more common. The new
> implementaiton will support them as well. I will just change the
> deprecation warnings.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/jasonlong/geo_pattern/pull/40#issuecomment-71384594>.
>



-- 
Thanks,
Taylor Blau

ttaylorr.com

maxmeyer commented:
> I disagree with it revealing too many implementation details

You're revealing the structure of your library in the API by using e.g. `GeoPattern::XesPattern`. If you want to refactor your library and you need to change the structure of your library you would change the API as well: `GeoPattern::XesPattern => GeoPattern::Generators::XesPattern`.

The reason I raise this issue is the following: When introducing the `Pattern`-class (#44) I think it does make sense to rename the `XesPattern` etc. as well and name them according their function, `XesPatternGenerator` or wrap them in a `Generator`-module.

maxmeyer commented on lib/geo_pattern/pattern_store.rb:
```diff
@@ -0,0 +1,55 @@
+    public
+
+    def initialize(
+      store = HashStore.new(
```
This is kind of a `HashWithIndifferentAccess`. So there's no need to use strings here anymore.

maxmeyer commented on lib/geo_pattern/pattern_store.rb:
```diff
@@ -0,0 +1,60 @@
+
+    def [](pattern)
+      if pattern.kind_of?(String) || pattern.kind_of?(Symbol)
+        $stderr.puts 'String pattern references are deprecated as of 1.3.0' if pattern.kind_of?(String)
```
I added the deprecation warnings for string + class. I can revert this change @jasonlong.

maxmeyer commented on spec/spec_helper.rb:
```diff
@@ -16,8 +16,7 @@
 # Loading support files
 #::File.expand_path('../support/*.rb', __FILE__)
-require_files_matching_pattern ::File.expand_path('../support/*.rb', __FILE__)
-require_files_matching_pattern ::File.expand_path('../shared_examples/*.rb', __FILE__)
```
Shared examples can now be found below `spec/support/shared_examples`.

maxmeyer commented on spec/support/helpers.rb:
```diff
@@ -1,32 +1,36 @@
-    b = (rgb.b * 255).round
-
-    format("rgb(%d, %d, %d)", r, g, b)
+  # Taken from active support
```
Borrowed from `active support` (MIT-licensed) to make deprecation warnings silent during testing. Otherwise they will litter the  `rspec` output.

maxmeyer commented:
@jasonlong @ttaylorr What do you think about the new structure?

ttaylorr commented on spec/support/helpers.rb:
```diff
@@ -1,32 +1,36 @@
+  #
+  # stream = capture(:stderr) { system('echo error 1>&2') }
+  # stream # => "error\n"
+  def capture(stream)
```
This is also in `Kernel::capture`?

tarebyte replied:
@ttaylorr it is, but this method is being deprecated due to it not being thread safe.

You can check it out here https://github.com/rails/rails/blob/346677b3137a35de65c660c230d6fe9e71aa9b0e/activesupport/lib/active_support/core_ext/kernel/reporting.rb

And here is the PR that talks about it https://github.com/rails/rails/pull/13392

Sorry for the random interjection I'm working a similar problem in `teachers_pet` :bowtie:

maxmeyer commented:
@ttaylorr Fixed. Yes, it's marked deprecated because of missing thread-safety, but I think this should be ok for testing. What do you think?

jasonlong commented:
Guys, I apologize for my responsiveness. I've been dealing with some personal/family matters lately and haven't been able to focus much on this. 

Most of this is lower level Ruby than I'm comfortable with, so I'm happy to defer to @ttaylorr, @tarebyte, and others to collaborate with you (@maxmeyer) on these decisions.

ttaylorr commented:
:+1:

maxmeyer commented:
:smile: Love to see this merge.

maxmeyer commented:
@jasonlong 
> Guys, I apologize for my responsiveness. I've been dealing with some personal/family matters lately and haven't been able to focus much on this. 

Hope everything is fine again for you/your familly soon!
